add make-unix-socket-tcp@

Bogdanp opened the pull request:
I used a similar implementation to improve the performance of the web server in the TechEmpower benchmarks and I figured it was worth adding here.

Bogdanp committed:
unix-socket-tcp-unit: add make-unix-socket-tcp@

Bogdanp commented:
👋 ping on this and #5.

Bogdanp commented:
Ping on this and #5 again. If this change seems too specific to the web-server to include here, let me know and I'll move it into its own package.

rmculpepper commented on unix-socket-lib/racket/unix-socket-tcp-unit.rkt:
```diff
@@ -0,0 +1,55 @@
+    (define (tcp-accept l)
+      (unix-socket-accept l))
+
+    (define (tcp-accept/enable-break l)
```
This implementation of `tcp-accept/enable-break` doesn't satisfy the guarantee that either the accept or the break occurs but not both. I think the following code should fix it (but I haven't tested it):

    (define (tcp-accept/enable-break l)
      (apply values (sync/enable-break (unix-socket-accept-evt l)))

Bogdanp replied:
Thanks! I misunderstood the docs of `tcp-accept/enable-break`. I'll fix that and try to add a test.

Bogdanp replied:
I made the change you suggested and I added a test that finds the problem with the old implementation under CS and BC if I add a loop like

```racket
  (time
   (let loop ([n 1000000000])
     (unless (zero? n)
       (loop (sub1 n)))))
```

to the top of `accept-poll/check` to increase the likelihood that the timing works out.

The new implementation never fails the test under BC, but it does fail it under CS. Looking through the documentation and at the implementation of `accept-evt`, I don't see any reason why this shouldn't work under CS as well so I think this could be some sort of bug in CS. I'll take another look at it tomorrow in case my test is broken and open an issue in the Racket repo if it's not.

Bogdanp replied:
That was indeed a Racket CS bug that got fixed in https://github.com/racket/racket/commit/4d0aa443b135d1d6fa3111ce934f63fe352aa60e . Since that commit, I can no longer reproduce the failure. Matthew added a similar test to the Racket test suite, so I went ahead and dropped mine from this repo.

Bogdanp committed:
tcp-unit: fix implementation of tcp-accept/enable-break

Bogdanp committed:
tests: add test for tcp-accept/enable-break

Bogdanp committed:
test: drop test specific to tcp-accept/enable-break

The Racket CS issue that this test found has been fixed in[1] and a
similar test has been added to the Racket test suite so this test no
longer has any long-term value.

[1]: https://github.com/racket/racket/commit/4d0aa443b135d1d6fa3111ce934f63fe352aa60e
